_posts/2019-05-20-server-rendering-responsively.md

```diff
@@ -0,0 +1,201 @@
+
+But with `react-responsive-media` as our primary approach, we felt that we could make some further optimizations
+with user agent detection. In the event that we don't know your device by its user agent, we'll still give you both
+the desktop and mobile breakpoints from SSR. But if we are certain you are on a mobile device, we can omit the
+desktop breakpoint from the server render.
+
+We really wanted to not maintain our own list of user agents. Alas, we found that none of the existing user agent
+detection libraries surfaced all the information we needed. We needed to know the minimum width for a browser on a
+given device, and if it was resizable, and to what dimensions it was resizable. If any existing libraries _did_
+have this data, they didn't provide it to us easily.
+
+So we did some experimentation, given the browsers and devices we knew we needed to support. And yeah...we
+(reluctantly) created our own user-agent detection library,
+[@artsy/detect-responsive-traits](https://github.com/artsy/detect-responsive-traits). We're using this to determine
+if your browser is likely going to use only the mobile breakpoint of our app, in which case we don't have to also
+render the desktop version. The library is currently targeting only the browsers and devices we support on
+[artsy.net](artsy.net), but
+[we're always open to contributions](https://github.com/artsy/detect-responsive-traits)!
+
+We aren't doing any detection of desktop browsers. Maybe this is something we do in the future, but for now we are
+more concerned with mobile users getting less content sent over their 3G connection.
+
+## Why didn't you use \_\_\_?
+
+Those are our primary tools for combining SSR with responsive design! They work well for us. We considered many
+many other options along the way. Here are a couple:
+
+### [`react-media`](https://github.com/ReactTraining/react-media) or [`react-responsive`]()
+
+We investigated both `react-media` and `react-responsive`, but found that they didn't approach the SSR side of the
+problem as deeply as we needed.
+
+We also weren't fans of the imperative API in `react-media`. We started with a similar API when building
+`react-responsive-media`, but learned that it limited our ability to render all breakpoints from the server. The
+API requires only one branch to be rendered, and this caused the aforementioned mismatches when the app would
+hydrate.
+
+With `react-responsive`, we didn't like that it relied on user agent detection as its primary method of handling
+SSR.
+
+### CSS
+
+We explored the idea of rendering all breakpoints from the server, and hiding the non-matching branches with CSS.
+The issue with this approach is that you have many components that are mounted and rendered unnecessarily. Aside
+from the performance hit you take for rendering components your user isn't seeing, even worse is the potential for
+duplicate side-effects.
```

Comment:
We actually _do_ do this and it works great. (I think that should actually be one of the first things to mention in the “How does it work” section, I'll write up a comment there.)

_However_, we couldn’t leave it at just that, because having React render **components** [as opposed to pure HTML/CSS] at _all_ breakpoints could lead to side-effects.

Reply:
I wrote up some suggestions to address this, but feel free to reword those.

Reply:
I think I've worked in enough changes that this is addressed...let me know if you feel otherwise!